contracts/external/traders/CanonicalOrders.sol

```diff
@@ -0,0 +1,842 @@
+/*
+
+    Copyright 2019 dYdX Trading Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.5.7;
+pragma experimental ABIEncoderV2;
+
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import { IAutoTrader } from "../../protocol/interfaces/IAutoTrader.sol";
+import { ICallee } from "../../protocol/interfaces/ICallee.sol";
+import { Account } from "../../protocol/lib/Account.sol";
+import { Math } from "../../protocol/lib/Math.sol";
+import { Monetary } from "../../protocol/lib/Monetary.sol";
+import { Require } from "../../protocol/lib/Require.sol";
+import { Types } from "../../protocol/lib/Types.sol";
+import { OnlySolo } from "../helpers/OnlySolo.sol";
+import { TypedSignature } from "../lib/TypedSignature.sol";
+
+
+/**
+ * @title CanonicalOrders
+ * @author dYdX
+ *
+ * Allows for Canonical Orders to be used with dYdX
+ */
+contract CanonicalOrders is
+    Ownable,
+    OnlySolo,
+    IAutoTrader,
+    ICallee
+{
+    using Math for uint256;
+    using SafeMath for uint256;
+    using Types for Types.Par;
+    using Types for Types.Wei;
+
+    // ============ Constants ============
+
+    bytes32 constant private FILE = "CanonicalOrders";
+
+    // EIP191 header for EIP712 prefix
+    bytes2 constant private EIP191_HEADER = 0x1901;
+
+    // EIP712 Domain Name value
+    string constant private EIP712_DOMAIN_NAME = "CanonicalOrders";
+
+    // EIP712 Domain Version value
+    string constant private EIP712_DOMAIN_VERSION = "1.0";
+
+    // Hash of the EIP712 Domain Separator Schema
+    /* solium-disable-next-line indentation */
+    bytes32 constant private EIP712_DOMAIN_SEPARATOR_SCHEMA_HASH = keccak256(abi.encodePacked(
+        "EIP712Domain(",
+        "string name,",
+        "string version,",
+        "uint256 chainId,",
+        "address verifyingContract",
+        ")"
+    ));
+
+    // Hash of the EIP712 CanonicalOrder struct
+    /* solium-disable-next-line indentation */
+    bytes32 constant private EIP712_ORDER_STRUCT_SCHEMA_HASH = keccak256(abi.encodePacked(
+        "CanonicalOrder(",
+        "bytes32 flags,",
```

Comment:
I can test, but it's going to be cheaper primarily because of data parsing and copying (sending to SignedOperationProxy, then SoloMargin, then CanonicalOrders several times) rather than saving on CALLDATA gas